### ISSUE TITLE
Mouse trajectories from deeplabcut as Input data 

### ISSUE THREAD

**Issue by Yaraat92** [kristinbranson/JAABA#326]
Hi, 

I would like to use JAABA to assess mouse grooming behavior. 
In our lab we use the open source software Deeplabcut (DLC) for tracking of mouse body parts, and all the input data required by JAABA and prepared by PrepareJAABAData GUI can be extracted using DLC. 
I was wondering how it would be possible to create adequate input for JAABA manually - not using the GUI.
Any suggestions?

Thanks in advance,
Yara.

**Comment by mkabra** [comment]
Hi Yara,

The documentation on how to prepare data for JAABA (
http://jaaba.sourceforge.net/DataFormatting.html) and the details of of the
trx file structure (
http://jaaba.sourceforge.net/DataFormatting.html#TracksFileStructure)
should enable you to write scripts that can prepare data for JAABA.

Mayank

**Comment by auesro** [comment]
Dear Yara,
I am trying something similar but for scratching and not using DLC at the moment cause I thought it was too cumbersome to calculate a and b...I was wondering how you are labeling the mice with DLC in order to obtain the JAABA parameters a and b....thanks a lot.
Cheers,
Augusto

**Comment by Yaraat92** [comment]
Hi Augusto,
That is also something I am still not sure about.. 
I was actually wondering whether 'a' and 'b' are parameters of the mouse? Do they refer to the long and short axes of the mouse's body?
If so then I might have an idea.. We track both ears of the mouse and the tail base (and the nose, but it is not always visible so it wouldn't be helpful for these analyses), and so you can draw a triangle connecting these 3 body parts and this should help calculate 'a' and 'b'.

**Comment by auesro** [comment]
Hi Yara,
Acording to JAABA documentation (http://jaaba.sourceforge.net/DataFormatting.html#TracksFileStructure), a is 1/4 lenght of the long axis of a ellipse fitted to the mouse while b is the same for the short axis. So in theory you would need to have both the lenght and the width of the mouse for every frame...unless Mayank contradicts me I dont know if that triangle would be enough for JAABA later to extract information....

**Comment by mkabra** [comment]
If you are unsure about what to use for 'a' and 'b', then set them to a
constant. The detectors that you train just won't use them.

**Comment by auesro** [comment]
But if you do that, from what I understand, you will be missing potential information that the detector could use to "learn" the behavior during training, right?

**Comment by mkabra** [comment]
Yeah, you would. The suggestion was in case there was no way of estimating
'a' and 'b'. If you can estimate them directly or indirectly then
definitely use the estimated values.

**Comment by auesro** [comment]
Thanks, thats what I expected.
And thats why I finally dropped the idea of using DLC...mayb Yara can try setting the points in head, tailbase, right flank and left flank...dont know how good DLC would learn tot differentiate right and left flanks but it might work...then calculating a and b would be straight forward.

**Comment by mkabra** [comment]
Well, in this case since you don't have exact values of 'a' and 'b', their
estimates or alternatives would also work. From detectors point of view,
'a' and 'b' are simply numbers representing aspects of the body's shape. In
fact, switching 'a' and 'b' also won't matter for the detector if you did
it consistently. So based on Yara's description, an alternative would be to
use the distance between the ears for 'b' and the distance between the
center of ears and the tail base as 'a'. In the worse case, if the
estimated values are not good, the detector won't use them.

**Comment by auesro** [comment]
Thats an option.
However, if b is the distance between ears, it will be pretty constant for mice since that doesnt change with the pose of the animal and so it would not add any important information from which to build a detector, right? But...now that I think of it...a and b for flies also dont change much...their bodies are pretty fixed by definition...  so how did you expect a and be to be useful for flies when you designed the code? Probably I am missing some understanding on how the code works.

**Comment by mkabra** [comment]
And just to add, we will soon be releasing Animal Part Tracker (APT)
software which implements DeepLabCut and other tracking algorithms. And APT
projects can directly be imported into JAABA, and JAABA will compute the
perframe features based on the landmarks. That way you won't have to figure
out how to convert tracked landmarks into JAABA compatible format.

**Comment by mkabra** [comment]
Well, they won't change within a fly, but they are quite different between
flies particularly between males and females. And we observed that when
flies groomed their wings, there were subtle changes in b, which were
enough to detect wing grooming even without tracking the wings. More
details are in the JAABA paper. So yes, they probably won't be helpful
directly but could be useful indirectly. So better to have them than not.

As for mice, 'b' won't be constant but would depend on the head's
direction, and as such could be a helpful signal.

**Comment by auesro** [comment]
Got it. For sure, I prefer to have it than not.

And thanks for the heads up regarding APT, looking forward to the release!

**Comment by Yaraat92** [comment]
Thanks for the all the input.. This all sounds very helpful.
I will try to use Mayank's suggestions for 'a' and 'b' from what I have for now, and see if it works well enough.. 
In any case, looking forward to using APT once released!

Yara.

**Comment by Yaraat92** [comment]
Hi again,

I noticed that the current version has an option to import from APT. Thus we have moved to work with APT, but I was wondering what animal parts I should track in a mouse so that 'a' and 'b' can be estimated correctly. 
As a trial run we started by marking both ears and the tailbase, in addition to right and left flank as suggested here.. Would that be enough for JAABA to convert these tracked parts to useful input?

**Comment by mkabra** [comment]
Hi Yara,

Yes, figuring out what points to track is difficult for mice which don't
have very clear landmarks. With the points you have selected JAABA will be
able to estimate the width ('a') and the length ('b'), but as you'll
probably realize that labeling the flanks won't be easy. Depending on your
video quality, you could try to track shoulders or thighs. Neck could be
another body part that might be worth tracking.

However, you should think more in terms of the behavior you would like to
detect. If you can tell me the behaviors that interest you, I might be able
to help you better. Feel free to message me privately if that's more
preferable to discuss the details.

Mayank

**Comment by Yaraat92** [comment]
Hi Mayank,

As I've mentioned in the beginning, I want to use JAABA to track events of grooming. 
It does sound helpful to have the 'a' and 'b' parameters. I am wondering whether we need to change our direction and try to use Motr for this purpose. What do you think?

**Comment by Yaraat92** [comment]
Hi,

Thanks for your detailed response.

I am actually following this paper: https://doi.org/10.1016/j.jneumeth.2017.05.026 . They suggest using Motr for tracking of the mouse and then training a JAABA classifier for detection of grooming behaviour. Since we had already used deeplabcut to track the mouse body parts I mentioned before, I was wondering whether I can use that as input instead of trajectories from Motr. 

I'm finding it a little too complicated to try and optimise what we get from DLC to fit the requirements of the JAABA classifier, and that's why I thought I might just try using Motr, since it seems to work well enough for this purpose.

Basically, these are the input parameters that are keeping me from being able to use DLC for my purpose:
1. theta - is it supposed to be in radians or degrees?
2. id - Identity number of the trajectory. What does this mean?
and the most important issue is:
3. 'b' - how to estimate b (the short axis of the mouse) using DLC.

After trying to use APT as suggested, I found it is not working well enough for us, and am trying to optimise its performance, but this is taking too long.. That's why now my plan is to try using what I have from DLC, entering 'b' as the distance between the ears and after I get answers about the issues above I will try to train a classifier and see whether it is successful. 
If not, I might consider using Motr as suggested in the paper....

**Comment by auesro** [comment]
Hi @Yaraat92 ,

Part of the info you need is published on the motr website, in particular [here](http://motr.janelia.org/motr_manual/looking_at_track_data.html#datafiles):
1. Radians.
2. ID of the mouse in case you have more than 1 mouse in the cage.
3. for b calculation you would need to label a point on the right and one point on the left side of the animals body at the level of the body center. Then just calculate as explained in motr: ` m_afB: length of semi-minor axis of ellipse in pixels `. I am pretty confident that DLC will learn to label those points accurately.

Good luck!

A

**Comment by Yaraat92** [comment]
Thanks a lot!! I will give a try :)

**Comment by mkabra** [comment]
Hi Yaara,

If you don't mind me asking, why isn't APT working as well for you? It has
DLC as one of the algorithms so it should perform as well. And if you are
worried about labeling again, we have a script in APT to import existing
DLC project:
https://github.com/kristinbranson/APT/blob/develop/matlab/createFromDeeplabcut.m

And also importing from APT projects is now available in JAABA. With this,
JAABA will compute all the relevant perframe features from the tracked data
and you don't have to manually convert the tracked data into JAABA
compatible format.

Mayank
